### ISSUE TITLE
Markdown lexer edge case with links and square brackets (rare)

### ISSUE THREAD

**Issue by Ravlen** [rouge-ruby/rouge#1444]
**Name of the lexer**
Markdown

This might be too rare to bother put effort into it, but this is something I run into from time to time as some of our engineers like linking to TOML section examples, and the section names are usually `[section]` or `[[section]]`.

I'm raising the issue because::

- It renders correctly with our markdown renderer (Kramdown): https://docs.gitlab.com/runner/configuration/autoscale.html#runner-global-options (see the two links in the table, which are correct).
- It does color properly with whatever code coloring VS Code uses:

<img width="1085" alt="Screen Shot 2020-02-20 at 14 16 12" src="https://user-images.githubusercontent.com/36979740/74903193-9aef1080-53eb-11ea-8e2b-ca337ac9e0fe.png">

**Code sample**

Does not work with GitHub's coloring either:

```markdown
[This is a link `with backticks`](example.com)
[This is a link to a TOML section `[with brackets]` (and backticks)](example.com)
[This is a link to a TOML section `[[with double brackets]]` (and backticks)](example.com)
```
Does not work with Rouge:

<img width="1167" alt="Screen Shot 2020-02-20 at 14 31 08" src="https://user-images.githubusercontent.com/36979740/74903886-a8a59580-53ed-11ea-8ac9-a7aaf1110c03.png">

Works fine with VS Code:

<img width="953" alt="Screen Shot 2020-02-20 at 14 31 51" src="https://user-images.githubusercontent.com/36979740/74903915-c246dd00-53ed-11ea-9384-5199b8d0d2ec.png">

**Additional context**
Yeah, it's a rare edge case. Now that I saw one fix earlier this week (that was fast, btw!), I'll see if I can spot any issues with the regex (though I'm still learning).

[dingus]: http://rouge.jneen.net/

**Comment by pyrmont** [comment]
Agreed this should parse. It works fine [in CommonMark](https://spec.commonmark.org/dingus/?text=%5BThis%20is%20a%20link%20%60with%20backticks%60%5D(example.com)%20%20%0A%5BThis%20is%20a%20link%20to%20a%20TOML%20section%20%60%5Bwith%20brackets%5D%60%20(and%20backticks)%5D(example.com)%20%20%0A%5BThis%20is%20a%20link%20to%20a%20TOML%20section%20%60%5B%5Bwith%20double%20brackets%5D%5D%60%20(and%20backticks)%5D(example.com)).

**Comment by pyrmont** [comment]
The problem is with how the regex uses non-greedy matching in rules like this:

https://github.com/rouge-ruby/rouge/blob/65207be96258926ef73da611a6a5c3e052019ca8/lib/rouge/lexers/markdown.rb#L105

The problem that I can imagine is that a more complex set of rules potentially leads you to a state where it's difficult to know on an initial pass whether you're in a link or not.

**Comment by Ravlen** [comment]
@pyrmont Would it be crazy to add the opening bracket for the link, as a way to identify the full link-text section?

```diff
- rule %r/(!?\[)(#{edot}*?)(\])/ do
+ rule %r/(!?\[)(#{edot}*?)(\]\()/ do
```

<img width="1086" alt="Screen Shot 2020-02-20 at 15 00 48" src="https://user-images.githubusercontent.com/36979740/74905572-d8ef3300-53f1-11ea-9d35-66d717f85c0e.png">
<img width="1086" alt="Screen Shot 2020-02-20 at 15 00 54" src="https://user-images.githubusercontent.com/36979740/74905582-df7daa80-53f1-11ea-8b7c-1e960e6b790d.png">

**Comment by pyrmont** [comment]
It would need to be either a `(` or a `[` (since you can have inline links and reference links) and it might be preferable to do it as a lookahead so that the rest of the lexing logic stays the same.

**Comment by Ravlen** [comment]
(Sidenote, I just noticed you and jneen are both in Japan, which is funny because I'm in Kagoshima!)

**Comment by Ravlen** [comment]
@pyrmont I struggle with lookahead logic and never got a good grasp of it, but I could easily add a `(` OR `[`:

```rb
rule %r/(!?\[)(#{edot}*?)(\]\(|\]\[)/ do
```

<img width="1086" alt="Screen Shot 2020-02-20 at 15 09 56" src="https://user-images.githubusercontent.com/36979740/74906017-17d1b880-53f3-11ea-8137-adf87e7f4c92.png">

**Comment by pyrmont** [comment]
You can think of a lookahead as a special kind of 'group' (using `(?=...)`). So in this case: `(!?\[)(#{edot}*?)(?=\][\[(])`. It's difficult to read because of all the escaping but in plain English, that's a `]` and then a character class consisting of either `[` or `(`.

**Comment by Ravlen** [comment]
@pyrmont I read through some regex docs and I'm starting to grasp it, but with the tool I was using your example wasn't working, though a slight modification worked. Does this make sense?

```ruby
rule %r/(!?\[)(#{edot}*?)(?=\]\(|\]\[)/ do
```

<img width="1086" alt="Screen Shot 2020-02-20 at 16 03 18" src="https://user-images.githubusercontent.com/36979740/74908946-8bc38f00-53fa-11ea-8cf4-0bb48148297e.png">

Once I got that working, I started to wonder if I should use lookbehind to skip the first bracket as well. Does this make sense?

```ruby
rule %r/(!?(?<=\[))(#{edot}*?)(?=\]\(|\]\[)/ do
```

<img width="1086" alt="Screen Shot 2020-02-20 at 16 03 05" src="https://user-images.githubusercontent.com/36979740/74908979-9c740500-53fa-11ea-8450-8d85d69f3942.png">
